buckets-mdapi/src/conditional.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright 2020 Joyent, Inc.
+
+use serde_json::Value;
+use postgres::types::ToSql;
+use postgres::Transaction;
+use serde_json;
+use slog::{trace, Logger};
+use uuid::Uuid;
+use serde_derive::{Deserialize, Serialize};
+
+use crate::object;
+use crate::error;
+use crate::metrics;
+use crate::sql;
+use crate::types;
+
+#[derive(Default, Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub struct Conditions {
+    #[serde(alias = "if-match")]
+    pub if_match: Option<Vec<String>>,
+
+    #[serde(alias = "if-none-match")]
+    pub if_none_match: Option<Vec<String>>,
+
+    #[serde(alias = "if-modified-since")]
+    pub if_modified_since: Option<types::Timestamptz>,
+
+    #[serde(alias = "if-unmodified-since")]
+    pub if_unmodified_since: Option<types::Timestamptz>,
+}
+
+pub fn error(error_type: error::BucketsMdapiErrorType, msg: String) -> serde_json::Value {
+    serde_json::to_value(error::BucketsMdapiError::with_message(
+        error_type,
+        msg,
+    ))
+    .expect("failed to encode a PreconditionFailedError error")
+}
+
+pub fn request(
+    mut txn: &mut Transaction,
+    items: &[&dyn ToSql],
+    vnode: u64,
+    conditions: &Conditions,
+    metrics: &metrics::RegisteredMetrics,
+    log: &Logger,
+) -> Result<(), Value> {
+    if !is_conditional(conditions) {
+        trace!(log, "request not conditional; returning");
+        return Ok(());
+    }
+
+    sql::txn_query(
+        sql::Method::ObjectGet,
+        &mut txn,
+        sql::get_sql(vnode).as_str(),
+        items,
+        metrics,
+        log,
+    )
+    .map_err(|e| { sql::postgres_error(e.to_string()) })
+    .and_then(|rows| {
+        if rows.is_empty() {
+            return Err(object::object_not_found());
+        } else if rows.len() > 1 {
+            return Err(sql::postgres_error("expected 1 row from conditional query".to_string()));
+        }
+
+        let object_id: Uuid = rows[0].get("id");
+        let object_modified: types::Timestamptz = rows[0].get("modified");
+        check_conditional(&conditions, object_id.to_string(), object_modified)?;
+
+        Ok(())
+    })
+}
+
+fn is_conditional(conditions: &Conditions) -> bool {
```

Comment:
For `is_conditional` and `check_conditional` you could consider making them methods attached to the `Conditions` type just to enable a call syntax of `conditions.is_conditional()` or `conditions.check()`. Doesn't affect functionality at all so feel free to ignore.

Reply:
I like it, thanks! `check` and `is_conditional` are now methods of the `Conditions` type.